tools/kagami/src/swarm.rs

```diff
@@ -0,0 +1,1314 @@
+                .relative_to(&self.target_dir)?
+                .to_str()
+                .wrap_err("config directory path is not a valid string")?
+                .to_owned(),
+            DIR_CONFIG_IN_DOCKER.to_owned(),
+        )];
+
+        let trusted_peers: BTreeSet<PeerId> =
+            peers.values().map(peer_generator::Peer::id).collect();
+
+        let mut peers_iter = peers.iter();
+
+        let first_peer_service = {
+            let (name, peer) = peers_iter.next().expect("There is non-zero count of peers");
+            let service = DockerComposeService::new(
+                peer,
+                service_source.clone(),
+                volumes.clone(),
+                trusted_peers.clone(),
+                Some(genesis_key_pair),
+            );
+
+            (name.clone(), service)
+        };
+
+        let services = peers_iter
+            .map(|(name, peer)| {
+                let service = DockerComposeService::new(
+                    peer,
+                    service_source.clone(),
+                    volumes.clone(),
+                    trusted_peers.clone(),
+                    None,
+                );
+
+                (name.clone(), service)
+            })
+            .chain(std::iter::once(first_peer_service))
+            .collect();
+
+        let compose = DockerCompose::new(services);
+        Ok(compose)
+    }
+}
+
+#[derive(Clone, Debug)]
+struct AbsolutePath {
+    path: PathBuf,
+}
+
+impl Deref for AbsolutePath {
+    type Target = PathBuf;
+
+    fn deref(&self) -> &Self::Target {
+        &self.path
+    }
+}
+
+impl AsRef<Path> for AbsolutePath {
+    fn as_ref(&self) -> &Path {
+        self.path.as_path()
+    }
+}
+
+impl AbsolutePath {
+    fn absolutize(path: PathBuf) -> Result<Self> {
+        Ok(Self {
+            path: if path.is_absolute() {
+                path
+            } else {
+                path.absolutize()?.to_path_buf()
+            },
+        })
+    }
+
+    /// Relative path from self to other.
+    fn relative_to(&self, other: &AbsolutePath) -> Result<PathBuf> {
+        pathdiff::diff_paths(self, other)
+                .ok_or_else(|| {
+                    eyre!(
+                        "failed to build relative path from {} to {}",
+                        other.display(),
+                        self.display(),
+                    )
+                })
+                // docker-compose might not like "test" path, but "./test" instead 
+                .map(|rel| {
+                    if rel.starts_with("..") {
+                        rel
+                    } else {
+                        Path::new("./").join(rel)
+
+                    }
+                })
+    }
+}
+
+mod key_gen {
+    use iroha_crypto::{error::Error, KeyGenConfiguration, KeyPair};
+
+    /// If there is no base seed, the additional one will be ignored
+    pub fn generate(base_seed: Option<&[u8]>, additional_seed: &[u8]) -> Result<KeyPair, Error> {
```

Comment:
Are we doing something different here from e.g. parent `kagami` module that it warrants a separate module?

Reply:
It might be just a function in `kagami::swarm` module, actually. It is used in a couple of places: in peers generation and in genesis key generation. 

There is also a key pair generation implementation in `kagami::crypto` module, though. That one also check an additional invariant I don't check here (32 bytes limit for secp). I should probably reuse that piece of code here.

Reply:
Refactored